File: Data_Structures/validParenthesis.java
import java.util.*;

class stack{
    private
        static final int MAX = 1000;
        int top;
        char[] st = new char[MAX];

    public
        //Constructor
        stack(){
            top = -1;
        }

        //push the character
        void push(char data){
            st[++top] = data;
        }

        //pop top most character
        void pop(){
            top--;
        }

        //return the element at the top of the stack
        char atTop(){
            return st[top];
        }

        //check whether the stack is empty
        boolean empty(){
            return top == -1;
        }

        //check whether the element at top of the stack and the incoming element are matching
        boolean match(char ch){
            boolean ans;
            if(st[top] == '(' && ch == ')'){
                ans = true;
            } else if(st[top] == '{' && ch == '}'){
                ans = true;
            } else if(st[top] == '[' && ch == ']'){
                ans = true;
            } else if(st[top] == '<' && ch == '>'){
                ans = true;
            } else{
                ans = false;
            }
            return ans;
        }
}

public class validParenthesis {
    public static void main(String[] args){

        stack s = new stack();

        Scanner scan = new Scanner(System.in);
        String str = scan.nextLine();

        for(int i = 0; i < str.length(); i++){
            if(!isParenthesis(str.charAt(i))){
                continue;
            }

            if(str.charAt(i) == '(' || str.charAt(i) == '{' || str.charAt(i) == '[' || str.charAt(i) == '<'){
                s.push(str.charAt(i));
            } else{
                if(s.empty()){
                    System.out.println("Invalid");
                    return;
                }else{
                    if(s.match(str.charAt(i))){
                        s.pop();
                    }
                    else{
                        System.out.println("Invalid");
                        return;
                    }
                }
            }
        }
        if(s.empty()){
            System.out.println("Valid");
        } else{
            System.out.println("Invalid");
        }
    }

    //check whether the incoming character is parenthesis or not
    static boolean isParenthesis(char ch){
        switch(ch){
            case '(':
            case '{':
            case '[':
            case '<':
            case ')':
            case '}':
            case ']':
            case '>':
                return true;
        }
        return false;
    }
}

File: Data_Structures/C/Stack/validParenthesis.c
#include<stdio.h>
#include<string.h>

#define MAX 100

void push(char stack[], int *top, char val){
    if(*top != MAX - 1){
        stack[++(*top)] = val;
    }
}

void pop(char stack[], int *top){
    if(*top != -1){
        (*top)--;
    }
}

int empty(int top){
    return top == -1;
}

int matching(char a, char b){
    if( (a == '(' && b == ')') || (a == '{' && b == '}') || (a == '[' && b == ']') || (a == '<' && b == '>')){
        return 1;
    }
    return 0;
}

int valid(char str[]){
    char stack[MAX];
    int top = -1;
    for(int i = 0; i<strlen(str); i++){
        
        // if(str[i] == ' '){
        //     continue;
        // }

        if(str[i] == '(' || str[i] == '{' || str[i] == '[' || str[i] == '<'){
            push(stack, &top, str[i]);
        }
        else if(str[i] == ')' || str[i] == '}' || str[i] == ']' || str[i] == '>'){
            if(empty(top)){
                return 0;
            }
            if(matching(stack[top], str[i])){
                pop(stack, &top);
            }
            else{
                return 0;
            }
        }
    }
    return empty(top);
}

int main(){
    char str[MAX];
    scanf("%[^\n]s",str);
    printf(valid(str) == 1 ? "Valid\n" : "Invalid\n");

    return 0;
}